core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs

```diff
@@ -0,0 +1,242 @@
+use std::{fmt::Debug, sync::Arc};
+
+use anyhow::Context;
+use async_trait::async_trait;
+use once_cell::sync::OnceCell;
+use tokio::{runtime::Handle, sync::watch};
+use zksync_dal::{ConnectionPool, StorageProcessor};
+use zksync_state::{
+    PostgresStorage, ReadStorage, RocksdbStorage, RocksdbStorageBuilder, StateKeeperColumnFamily,
+};
+use zksync_storage::RocksDB;
+use zksync_types::{L1BatchNumber, MiniblockNumber};
+
+/// Factory that can produce a [`ReadStorage`] implementation on demand.
+#[async_trait]
+pub trait ReadStorageFactory: Debug + Send + Sync + 'static {
+    async fn access_storage(
```

Comment:
I wonder whether it would make sense to have a more defensive interface that takes the executed L1 batch number as an argument. Could be dealt with later, I suppose.

Reply:
Hmm, defensive in the sense that it would sanity check that the current state of PG matches the latest executed L1 batch or in a sense that it would force `AsyncRocksdbCache` to cut off state at that specific L1 batch somehow?

Reply:
The first option, i.e. a sanity check.